internet-shop/src/main/java/com/andersen/shop/ProgramEngine.java

```diff
@@ -33,25 +63,137 @@ public static void run(Catalog catalog) {
                     continue;
                 case 5:
                     basket.clear();
+                    continue;
+                case 6:
+                    serializeBasket();
+                    exit();
             }
         }
     }
 
-    public static void addToBasket(Catalog catalog) {
-        int quantity;
+    private static void warehouseProcessing(Catalog catalog) {
+        while (true) {
+            System.out.println("\n\n 1) Add product 2) Display products");
+            int selection = input.nextInt();
+            switch (selection) {
+                case 1:
+                    addToWarehouse(catalog);
+                    break;
+                case 2:
+                    displayProductsFromWarehouse(catalog);
+                    break;
+            }
+        }
+    }
+
+    private static void displayProductsFromCatalog(Catalog catalog) {
+        Map<Integer, Product> products = catalog.getProducts();
+        for (Integer integer : products.keySet()) {
+            Product product = products.get(integer);
+            System.out.printf("%-4s %-10s %-6s %-3s%n", product.getId(), product.getName(),
+                    countPrice(product.getPrice(), product.getCurrency().getCode()), "rub");
+        }
+    }
 
+    private static void displayProductsFromWarehouse(Catalog catalog) {
+        Map<Integer, Integer> items = warehouse.getItems();
+        Set<Integer> itemsId = items.keySet();
+        for (Integer itemId : itemsId) {
+            Product product = catalog.getProduct(itemId);
+            if (product instanceof ProductWithExpiryDate) {
+                System.out.printf("%-4s %-10s %-3s %-6s %-20s%n", product.getId(), product.getName(), items.get(itemId), "pcs", ((ProductWithExpiryDate) product).getExpiryDate());
+            } else {
+                System.out.printf("%-4s %-10s %-3s %-6s%n", product.getId(), product.getName(), items.get(itemId), "pcs");
```

Comment:
You use the redundant operations.
items are gotten a key set and then you are searching in the original map the value.
use Map.Entry<Integer, Integer> entry : items